### ISSUE TITLE
Pure virtual 

### ISSUE THREAD

**Issue by aheejin** [emscripten-core/emscripten#17339]

**Comment by sbc100** [comment]
Sadly I don't think that would be easy to fix since clang hardcodes a single named for this function:

https://github.com/llvm/llvm-project/blob/feb44ccc913b3de6ec5ab9f74011e99614991439/clang/lib/CodeGen/ItaniumCXXABI.cpp#L308

**Comment by xbcnn** [comment]
The foo() call in Parent::Parent() ( `ctor` ) shouldn't be statically resolved and compiling error?

**Comment by sbc100** [comment]
Clearly that it not the case otherwise this example would not compile, which it does.

(I imagine that in this simple case you could tell that `foo()` was called from the constructor, but in the general case (where the constructor call any other function that then calls back into the virtual function) this would be hard/impossible for the compiler to verify, so the decided not to specify this as an error).

**Comment by aheejin** [comment]
@sbc100 
> Sadly I don't think that would be easy to fix since clang hardcodes a single named for this function:
> 
> https://github.com/llvm/llvm-project/blob/feb44ccc913b3de6ec5ab9f74011e99614991439/clang/lib/CodeGen/ItaniumCXXABI.cpp#L308

We can override that function, which is virtual.

**Comment by sbc100** [comment]
> @sbc100
> 
> > Sadly I don't think that would be easy to fix since clang hardcodes a single named for this function:
> > https://github.com/llvm/llvm-project/blob/feb44ccc913b3de6ec5ab9f74011e99614991439/clang/lib/CodeGen/ItaniumCXXABI.cpp#L308
> 
> We can override that function, which is virtual.

That doesn't help because it can only return a single value, and we would need to somehow return N different function names, one per signature.   I'm not saying its not possible, but it looks very non-trivial.

**Comment by aheejin** [comment]
Hmm, you're right. One option is to change `GetPureVirtualCallName()` to take an additional `NamedDecl` or something that contains info of the signature, and set its default parameter value to `void(void)`. This requires changes in the parent class, so it's possible someone might object to it.

Another option is fixing that up within our backend, as we do for invokes, with which we convert names like `__invoke_void_int` to `__invoke_vi`.

**Comment by sbc100** [comment]
Agreed that both of those are possible options.  I'm not sure its worth the effort though?   Do you think this is common rnough problem?

One issue is that ff we make the `__cxa_pure_virtual` function symbol names unbounded/unlimited then it makes it hard for them to be implement in native code and they will probably end up being auto-generated as JS imports, which adds some complexity (and size) I think.

**Comment by aheejin** [comment]
Yeah it will increase the code size and complexity. That's the reason I said it may only be justified in debug+assert builds. On how common this problem is, I don't know, but I guess not very common to accidentally write that kind of code. I don't think this is a priority for us to fix in near future, but filed the issue just not to completely forget about it.
